Dev/add error code (#445)
quicwg/multipath: draft-ietf-quic-multipath.md

  Receipt of a value that is higher than the local maximum value MUST
  be treated as a connection error of type PROTOCOL_VIOLATION.

# Error Codes {#error-codes}

> Multipath QUIC transport error codes are 62-bit unsigned integers

Author: ```suggestion
QUIC transport error codes are 62-bit unsigned integers
```

following {{QUIC-TRANSPORT}}.

> The following error codes are defined for use in the PATH_ABANDON frame.

Reviewer: ```suggestion
The following additional QUIC error codes are defined for use in the PATH_ABANDON frame.
```
I assume the intention is to be able also use e.g. NO_ERROR, right? Maybe it's good to say that explicitly?

> The following additional QUIC error codes are defined for use in the PATH_ABANDON frame.

Author: ```suggestion
In addition to NO_ERROR(0x0) following {{QUIC-TRANSPORT}}, the following QUIC error codes are defined for use in the PATH_ABANDON frame.
```

Reviewer 2: While I agree that NO_ERROR is a fine error message for a PATH_ABANDON, the current text is not really clear about the other error codes defined in RFC9000. For instance, what should a receiver do if it receives a PATH_ABANDON with error code FLOW_CONTROL_ERROR? This would sound wrong to me, though it is not clear how a receiver would react.

Reviewer: I don't think the receiver would react it all. This is only for logging. If there are concerns about this that I think we would need to define a separate code space. Thinking more about this, I think using the QUIC error code space is wrong. I think we should not call this error code at all. In case of connection close, you actually close the connection because there was an unexpected protocol behavior. In case of the multipath extension, it is the same: if there is something unexpected, you have to close the whole connection. Therefore closing a path is never an error but always an explicit decision.

Reviewer 2: Yes, I agree with your point. If we need such logging, we may standardise a few PATH_ABANDON-specific reason code points (the three mentioned above seem sufficient).

Reviewer 3: As a reference, [HTTP/3's error code space](https://www.iana.org/assignments/http3-parameters/http3-parameters.xhtml#http3-parameters-error-codes) contains codes that cover various scenarios.

This include unexpected protocol errors, such as detecting the peer closed a stream it shouldn't and responding with a CONNECTION_CLOSE of H3_CLOSED_CRITICAL_STREAM.

But it also includes expected behaviours that are not errors, such as when an endpoint cancels a request (a bit like path abandon) by sending a REST_STREAM or STOP_SENDING using H3_REQUEST_CANCELLED.

In all these cases, an H3 implementation has space to ignore the specific language in the spec: use the wrong error code space (send a transport error when it was an application error), or send a generic code, or send an incorrect code. I don't think any of those cause any real problems for the recipient, as Mirja put it, it's mainly for debugging. And someone investigating these should always be skeptical of the peer not telling the ground truth.

On balance, I think its fine to use the QUIC error code space here. This is an example of an extension where we are building some more complexity into the transport and can expect to need to extend the shared codepoint space to suit.

Reviewer 4: I also think that using the QUIC error space is fine. It is fairly obvious that not all error codes defined in the QUIC error space apply to all scenarios. For example, a "transport parameter error" only makes sense when parsing transport parameters. The important property is that the error codes be unique.

I also do not think that any implementation should react to specific error code values, unless that's explicitly mentioned in the text of the standard. The purpose is documentation for helping debugging issues, or for passing information to the application and ultimately the user. We don't need a special registry for path abandon reasons.

> In addition to NO_ERROR(0x0) following {{QUIC-TRANSPORT}}, the following QUIC error codes are defined for use in the PATH_ABANDON frame.

Reviewer: ```suggestion
In addition to NO_ERROR(0x0) (see {{Section 21.1 of QUIC-TRANSPORT}}, the following QUIC error codes are defined for use in the PATH_ABANDON frame.
```

> QUIC transport error codes are 62-bit unsigned integers
> following {{QUIC-TRANSPORT}}.
> 
> In addition to NO_ERROR(0x0) following {{QUIC-TRANSPORT}}, the following QUIC error codes are defined for use in the PATH_ABANDON frame.

Reviewer: ```suggestion
QUIC transport error codes are 62-bit unsigned integers (see {{Section 20.1 of QUIC-TRANSPORT}}.
In addition to NO_ERROR(0x0), the following QUIC error codes are defined
for use in the PATH_ABANDON frame:
```

> APPLICATION_ABANDON (TBD): The endpoint is abandoning the path at the

Reviewer 5: ```suggestion
APPLICATION_ABANDON (TBD-09): The endpoint is abandoning the path at the
```

request of the application. The application has determined that it no
longer needs this path. This error is used when the application layer
decides to stop using a specific path.

> RESOURCE_LIMIT_REACHED (TBD): The endpoint is abandoning the path because

Reviewer 5: ```suggestion
RESOURCE_LIMIT_REACHED (TBD-10): The endpoint is abandoning the path because
```

QUIC transport error codes are 62-bit unsigned integers
following {{QUIC-TRANSPORT}}.

In addition to NO_ERROR(0x0) following {{QUIC-TRANSPORT}}, the following QUIC error codes are defined for use in the PATH_ABANDON frame.

> 

Reviewer 4: Should we add provisional values for the two error codes in the IANA section, to enable interop testing?

Reviewer 5: I support @huitema 's idea, we need two experimental error codes value here for interop testing.

TBD-08 (experiments use 0x15228c0d)                  | PATHS_BLOCKED    | {{paths-blocked-frame}}
{: #frame-types title="Addition to QUIC Frame Types Entries"}

The following transport error code defined in {{tab-error-code}} are to
be added to the "QUIC Transport Error Codes" registry under
the "QUIC Protocol" heading.

Value                       | Code                  | Description                   | Specification
----------------------------|-----------------------|-------------------------------|-------------------

> TBD | APPLICATION_ABANDON | Path abandoned at the application's request | {{error-codes}}

Reviewer 5: ```suggestion
TBD-09(experiments use 0x4150504C4142414E) | APPLICATION_ABANDON | Path abandoned at the application's request | {{error-codes}}
```

> TBD | RESOURCE_LIMIT_REACHED | Path abandoned due to resource limitations in the transport | {{error-codes}}

Reviewer 5: ```suggestion
TBD-10(experiments use 0x5245534C494D4954) | RESOURCE_LIMIT_REACHED | Path abandoned due to resource limitations in the transport | {{error-codes}}
```
